File: src/citeguard/crossref.py
"""Minimal Crossref REST API client.

Uses only ``urllib`` from the standard library -- no ``requests``
dependency. Crossref's API is free and requires no API key; including a
``mailto`` in the User-Agent opts a caller into Crossref's faster "polite
pool" (see https://api.crossref.org), which this client does by default
when a contact email is configured.
"""

from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Callable

CROSSREF_BASE_URL = "https://api.crossref.org/works"
DEFAULT_USER_AGENT = "citeguard/1.0 (https://github.com/wedo911/citeguard)"


class DoiNotFoundError(Exception):
    """The DOI does not exist in Crossref."""


class CrossrefError(Exception):
    """Any other Crossref API failure (network error, unexpected status, malformed response)."""


Fetcher = Callable[[str], dict]


def build_fetcher(*, contact_email: str | None = None, timeout: float = 10.0) -> Fetcher:
    """Build a fetch function: doi -> Crossref ``message`` dict.

    Kept as a factory (rather than a single hardcoded function) so tests
    can inject a fake fetcher instead of hitting the network, and so
    callers can configure the polite-pool contact email and timeout.
    """
    user_agent = DEFAULT_USER_AGENT
    if contact_email:
        user_agent = f"{DEFAULT_USER_AGENT} (mailto:{contact_email})"

    def fetch(doi: str) -> dict:
        url = f"{CROSSREF_BASE_URL}/{urllib.parse.quote(doi, safe='')}"
        request = urllib.request.Request(url, headers={"User-Agent": user_agent})
        try:
            with urllib.request.urlopen(request, timeout=timeout) as response:
                body = response.read()
        except urllib.error.HTTPError as exc:
            if exc.code == 404:
                raise DoiNotFoundError(doi) from exc
            raise CrossrefError(f"Crossref returned HTTP {exc.code} for {doi}") from exc
        except urllib.error.URLError as exc:
            raise CrossrefError(f"Network error contacting Crossref for {doi}: {exc.reason}") from exc

        try:
            payload = json.loads(body)
        except json.JSONDecodeError as exc:
            raise CrossrefError(f"Crossref returned malformed JSON for {doi}") from exc

        message = payload.get("message")
        if not isinstance(message, dict):
            raise CrossrefError(f"Crossref response for {doi} had no 'message' object")
        return message

    return fetch


def polite_fetcher(fetcher: Fetcher, *, delay_seconds: float = 0.1) -> Fetcher:
    """Wrap a fetcher with a small fixed delay between calls, so checking a
    whole bibliography doesn't hammer Crossref's API in a tight loop.
    """

    def wrapped(doi: str) -> dict:
        result = fetcher(doi)
        time.sleep(delay_seconds)
        return result

    return wrapped

File: mcp-server/src/services/crossref.ts
/**
 * Minimal Crossref REST API client using the native `fetch` (Node 18+) --
 * no HTTP library dependency. Mirrors ../../../src/citeguard/crossref.py.
 */

const CROSSREF_BASE_URL = "https://api.crossref.org/works";
const DEFAULT_USER_AGENT = "citeguard-mcp-server/1.0 (https://github.com/wedo911/citeguard)";

export class DoiNotFoundError extends Error {
  constructor(doi: string) {
    super(`DOI not found in Crossref: ${doi}`);
    this.name = "DoiNotFoundError";
  }
}

export class CrossrefError extends Error {}

export type Fetcher = (doi: string) => Promise<Record<string, unknown>>;

export function buildFetcher(options: { contactEmail?: string; timeoutMs?: number } = {}): Fetcher {
  const userAgent = options.contactEmail ? `${DEFAULT_USER_AGENT} (mailto:${options.contactEmail})` : DEFAULT_USER_AGENT;
  const timeoutMs = options.timeoutMs ?? 10000;

  return async function fetchWork(doi: string): Promise<Record<string, unknown>> {
    const url = `${CROSSREF_BASE_URL}/${encodeURIComponent(doi)}`;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);

    let response: Response;
    try {
      response = await fetch(url, { headers: { "User-Agent": userAgent }, signal: controller.signal });
    } catch (err) {
      throw new CrossrefError(
        `Network error contacting Crossref for ${doi}: ${(err as Error).message}. ` +
          "If this happens for every DOI, your MCP client may be spawning this server with a " +
          "minimal environment that drops a variable needed for outbound HTTPS in your network " +
          "(e.g. NODE_EXTRA_CA_CERTS for corporate/antivirus TLS inspection, or HTTP(S)_PROXY) -- " +
          "many MCP clients only pass a small default env allowlist to child processes rather than " +
          "the full parent environment. Try configuring your client to inherit the full environment, " +
          "or explicitly pass the specific variable your network requires. See the README's " +
          "Troubleshooting section."
      );
    } finally {
      clearTimeout(timeout);
    }

    if (response.status === 404) {
      throw new DoiNotFoundError(doi);
    }
    if (!response.ok) {
      throw new CrossrefError(`Crossref returned HTTP ${response.status} for ${doi}`);
    }

    let payload: unknown;
    try {
      payload = await response.json();
    } catch {
      throw new CrossrefError(`Crossref returned malformed JSON for ${doi}`);
    }

    const message = (payload as { message?: unknown })?.message;
    if (!message || typeof message !== "object") {
      throw new CrossrefError(`Crossref response for ${doi} had no 'message' object`);
    }
    return message as Record<string, unknown>;
  };
}

/** Wrap a fetcher with a small fixed delay so checking many DOIs doesn't hammer Crossref. */
export function politeFetcher(fetcher: Fetcher, delayMs = 100): Fetcher {
  return async (doi: string) => {
    const result = await fetcher(doi);
    await new Promise((resolve) => setTimeout(resolve, delayMs));
    return result;
  };
}
